src/Microsoft.Health.Fhir.Api/Features/Throttling/ThrottlingMiddleware.cs

```diff
@@ -55,6 +67,50 @@ public class ThrottlingMiddleware
                     _excludedEndpoints.Add((excludedEndpoint.Method, excludedEndpoint.Path));
                 }
             }
+
+#pragma warning disable 4014 // Intentional fire and forget behavior
+            SamplingLoop();
+#pragma warning restore 4014
+        }
+
+        public void Dispose() => _disposed = true;
+
+        /// <summary>
+        /// Samples the success rate (i.e. requests not throttled) over a period, and adjusts the value we return as the retry after header.
+        /// This is an extremely simple approach that exponentially grows or decays the value depending on whether the success rate is
+        /// less than or greater than 99%, respectively.
+        /// The value as high as necessary to keep the success rate over 99%, but we want it to decrease when the request rate backs off,
+        /// so overall latency for clients is not unnecessarily high.
+        /// </summary>
+        public async Task SamplingLoop()
+        {
+            while (!_disposed)
+            {
+                await Task.Delay(SamplePeriodMilliseconds);
+
+                var successCount = Interlocked.Exchange(ref _currentPeriodSuccessCount, 0);
+                var failureCount = Interlocked.Exchange(ref _currentPeriodRejectedCount, 0);
+
+                var totalCount = successCount + failureCount;
+                double successRate = totalCount == 0 ? 100.0 : successCount * 100.0 / totalCount;
+
+                if (successRate >= 99)
+                {
+                    // see if we should lower the value
+                    if (_currentRetryAfterMilliseconds > MinRetryAfterMilliseconds)
+                    {
+                        _currentRetryAfterMilliseconds = (int)(_currentRetryAfterMilliseconds / RetryAfterDecayRate);
```

Comment:
Nit: This could end up with a value below the MinRetryAfterMilliseconds. With the current values it couldn't go below 19ms though.

Reply:
Yeah, it's not super important that we stay exactly within these limits, more that the header doesn't reach some outrageously large number or less than 3, where because of truncation it will never grow.
But I can cap the results so they never surpass the min/max

Reply:
Capped in https://github.com/microsoft/fhir-server/pull/1610/commits/30820bc71b481457f431941e342ad86aa61634b4